src/nexusrpc/handler/_decorators.py

```diff
@@ -0,0 +1,341 @@
+
+    if cls is None:
+        return decorator
+
+    return decorator(cls)
+
+
+OperationHandlerFactoryT = TypeVar(
+    "OperationHandlerFactoryT", bound=Callable[[Any], OperationHandler[Any, Any]]
+)
+
+
+@overload
+def operation_handler(
+    method: OperationHandlerFactoryT,
+) -> OperationHandlerFactoryT: ...
+
+
+@overload
+def operation_handler(
+    *, name: Optional[str] = None
+) -> Callable[[OperationHandlerFactoryT], OperationHandlerFactoryT]: ...
+
+
+def operation_handler(
+    method: Optional[OperationHandlerFactoryT] = None,
+    *,
+    name: Optional[str] = None,
+) -> Union[
+    OperationHandlerFactoryT,
+    Callable[[OperationHandlerFactoryT], OperationHandlerFactoryT],
+]:
+    """
+    Decorator marking an operation handler factory method in a service handler class.
+
+    An operation handler factory method is a method that takes no arguments other than
+    `self` and returns an :py:class:`OperationHandler` instance.
+
+    Args:
+        method: The method to decorate.
+        name: Optional name for the operation. If not provided, the method name will be used.
+
+    Examples:
+        .. code-block:: python
+
+            @nexusrpc.handler.operation_handler
+            def my_operation(self) -> Operation[MyInput, MyOutput]:
+                ...
+
+        .. code-block:: python
+
+            @nexusrpc.handler.operation_handler(name="my-operation")
+            def my_operation(self) -> Operation[MyInput, MyOutput]:
+                ...
+    """
+
+    def decorator(
+        method: OperationHandlerFactoryT,
+    ) -> OperationHandlerFactoryT:
+        # Extract input and output types from the return type annotation
+        input_type = None
+        output_type = None
+
+        return_type = typing.get_type_hints(method).get("return")
+        if typing.get_origin(return_type) == OperationHandler:
+            type_args = typing.get_args(return_type)
+            if len(type_args) == 2:
+                input_type, output_type = type_args
+            else:
+                warnings.warn(
+                    f"OperationHandler return type should have two type parameters (input and output type), "
+                    f"but operation {method.__name__} has {len(type_args)} type parameters: {type_args}"
+                )
+
+        set_operation_definition(
+            method,
+            nexusrpc.Operation(
+                name=name or method.__name__,
+                method_name=method.__name__,
+                input_type=input_type,
+                output_type=output_type,
+            ),
+        )
+        return method
+
+    if method is None:
+        return decorator
+
+    return decorator(method)
+
+
+@overload
+def sync_operation(
+    start: Callable[
+        [ServiceHandlerT, StartOperationContext, InputT],
+        Union[OutputT, Awaitable[OutputT]],
+    ],
+) -> Callable[
+    [ServiceHandlerT, StartOperationContext, InputT],
+    Union[OutputT, Awaitable[OutputT]],
+]: ...
```

Comment:
Supporting both `async def` and `def` decorated functions in a single type signature has undesirable results:

<img width="791" alt="image" src="https://github.com/user-attachments/assets/cbb3035d-2b3a-4f6a-b014-9f6f38165762" />
 
I'm going to create `nexusrpc.handler.syncio` and house the syncio versions of `sync_operation` and `SyncOperationHandler` in there.